### PR TITLE
Subtyping relationship among polymorphic qualifiers

### PR DESCRIPTION
**Pull request by t-rasmud**

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java [hunk 1]
```diff
@@ -80,25 +83,57 @@ protected void replace(
...
                 return;
             case "use":
-                if (!replacements.contains(factory.POLYDET)
-                        && !replacements.contains(factory.POLYDET_UP)
-                        && !replacements.contains(factory.POLYDET_DOWN)
-                        && !replacements.contains(factory.POLYDET_UPDET)) {
-                    type.replaceAnnotations(replacements);
+                if (replacementsPolyDet != null
+                        && replacementsPolyDet.contains(factory.POLYDET_NOORDERNONDET)) {
+                    if (AnnotationUtils.containsSame(
```
Comment (smillst) [inline checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:97]: Isn't this condition always true given the outer if statement?

File: checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java [hunk 2]
```diff
@@ -80,25 +83,57 @@ protected void replace(
...
+                            replacementsPolyDet, factory.POLYDET_NOORDERNONDET)) {
+                        type.replaceAnnotation(factory.POLYDET_USENOORDERNONDET);
+                    }
+                }
+                if (replacementsPolyDet != null) {
```
Comment (smillst) [inline checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:102]: I don't understand what this is doing/trying to do.  Can you add a comment?

File: docs/manual/contributors.tex [hunk 3]
```diff
@@ -48,6 +48,7 @@
 Kartikeya Goswami,
 Kivanc Muslu,
 Konstantin Weitz,
+L. Sun,
```
Comment (smillst) [inline docs/manual/contributors.tex:51]: I think these changes can be reverted?

File: checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java [hunk 4]
```diff
@@ -80,25 +83,58 @@ protected void replace(
...
-                if (!replacements.contains(factory.POLYDET)
-                        && !replacements.contains(factory.POLYDET_UP)
-                        && !replacements.contains(factory.POLYDET_DOWN)
-                        && !replacements.contains(factory.POLYDET_UPDET)) {
-                    type.replaceAnnotations(replacements);
+                if (replacementsPolyDet != null
+                        && replacementsPolyDet.contains(factory.POLYDET_NOORDERNONDET)) {
+                    type.replaceAnnotation(factory.POLYDET_USENOORDERNONDET);
+                }
+                // If @PolyDet("use") resolves to a qualifier that is a subtype of
```
Comment (smillst) [inline checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:99]: This comment is the English equivalent of the code.  What's "type annotation"? What is in `replacementsPolyDetUse`?

File: checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java [hunk 5]
```diff
@@ -80,25 +83,65 @@ protected void replace(
...
-                        && !replacements.contains(factory.POLYDET_UP)
-                        && !replacements.contains(factory.POLYDET_DOWN)
-                        && !replacements.contains(factory.POLYDET_UPDET)) {
-                    type.replaceAnnotations(replacements);
+                if (replacementsPolyDet != null
+                        && replacementsPolyDet.contains(factory.POLYDET_NOORDERNONDET)) {
+                    type.replaceAnnotation(factory.POLYDET_USENOORDERNONDET);
+                }
+                // Replace @PolyDet("use") with @PolyDet if @PolyDet("use") doesn't
+                // resolve to a type that is a subtype of what @PolyDet resolves to.
```
Comment (smillst) [inline checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:100]: I thought `@PolyDet("use")` was suppose to resolve to the same annotation as `@PolyDet`.  Why is this only true if it is not a subtype?

Comment (t-rasmud) [reply checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:100]: I had changed the semantics of `@PolyDet("use")` (after making collections invariant) as follows:
if the type that `@PolyDet(“use”)` resolves to is a subtype of the type that `@PolyDet `resolves to
then do nothing
else replace `@PolyDet(“use”)` with the type that `@PolyDet` resolves to.

Example:
`List.addAll()` is annotated as:
`@PolyDet("down") boolean addAll(@PolyDet List<@PolyDet("use") E> this, @PolyDet("use") Collection<? extends E> c);`

Consider the following example:
```
void test(
       @OrderNonDet List<@Det Integer> lst,
       @Det List<@Det Integer> lst1,
       @OrderNonDet List<@Det Integer> lst2) {
   lst.addAll(lst1);
   lst.addAll(lst2);
}
```

The above code is sound and type checks with this implementation of `@PolyDet(“use”)`
In method call `lst.addAll(lst1)`, `@PolyDet(“use”)` resolves to `@Det` which is a subtype of what `@PolyDet` resolves to (`@OrderNonDet`). Similarly for the method call `lst.addAll(lst2);`

If we add a call `lst.addAll(lst3)` where `lst3` is of the type `@NonDet List<@Det Integer>`, then `@PolyDet(“use”) `resolves to `@OrderNonDet` and the checker reports an error.

Comment (t-rasmud) [reply checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:100]: Thanks, I have updated the manual. 
Note: the manual doesn't talk about invariant collections. I have left todos for future me and will update those sections when I create relevant PRs.

File: docs/manual/determinism-checker.tex [hunk 6]
```diff
@@ -238,6 +238,7 @@
 % }
 % \end{verbatim}
 
+%TODO: Rashmi: remove this
```
Comment (smillst) [inline docs/manual/determinism-checker.tex:241]: TODO

File: docs/manual/determinism-checker.tex [hunk 7]
```diff
@@ -432,6 +433,33 @@
 E set(@PolyDet List<E> this, @PolyDet("use") int index, E element)
 \end{Verbatim}
 
+As alluded to earlier, we allow \<@PolyDet("use")> to resolve to any
```
Comment (smillst) [inline docs/manual/determinism-checker.tex:436]: It's not _any_ qualifier. `@PolyDet("use")` is resolved as a separate polymorphic qualifier. If that qualifier is a supertype of the instantiation of `@PolyDet`, then the instantiation of `@PolyDet` is used as the instantiation of `@PolyDet("use")`.

File: docs/manual/determinism-checker.tex (hunk 6 above)
Comment (smillst) [reply docs/manual/determinism-checker.tex:241]: Adding unrelated TODO's in this pull request is distracting.

File: checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java [hunk 8]
```diff
@@ -80,25 +83,56 @@ protected void replace(
...
                 return;
             case "use":
-                if (!replacements.contains(factory.POLYDET)
-                        && !replacements.contains(factory.POLYDET_UP)
-                        && !replacements.contains(factory.POLYDET_DOWN)
-                        && !replacements.contains(factory.POLYDET_UPDET)) {
-                    type.replaceAnnotations(replacements);
+                if (replacementsPolyDet != null
+                        && replacementsPolyDet.contains(factory.POLYDET_NOORDERNONDET)) {
+                    type.replaceAnnotation(factory.POLYDET_USENOORDERNONDET);
```
Comment (smillst) [inline checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:97]: I looks like this annotation is replaced again in the next if statement.  So what is the purpose of this?

File: checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java [hunk 9]
```diff
@@ -114,16 +118,62 @@ protected void replace(
                         || replacements.contains(factory.ORDERNONDET)) {
                     type.replaceAnnotations(replacements);
                 } else if (replacements.contains(factory.DET)) {
-                    replaceForPolyWithModifier(type, factory.ORDERNONDET);
```
[the remark this answers is not in the document]
Comment (t-rasmud) [reply checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:117]: The call to `replaceForPolyWithModifier` in the `"down"` case is intentional. 
Consider the method `descendingSet` in `TreeSet` which is annotated as:
`public @PolyDet("down") NavigableSet<E> descendingSet(@PolyDet TreeSet<E> this`

If we do not replace all the nested types (by calling `replaceForPolyWithModifier`),
a call to `descendingSet` on a receiver of type `@OrderNonDet TreeSet<@OrderNonDet TreeSet<@Det String>>` will result in an invalid type `@Det TreeSet<@OrderNonDet TreeSet<@Det String>>`.

[the remark this answers is not in the document]
Comment (smillst) [reply checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:117]: If the receiver has type  @OrderNonDet TreeSet<@OrderNonDet HashSet<@Det String>> why is it sound to call @Det TreeSet< @Det HashSet<@Det String>>?

[the remark this answers is not in the document]
Comment (t-rasmud) [reply checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java:117]: As discussed in our last meeting, I have changed this to `type.replaceAnnotation` after invalidating `@OrderNonDet TreeSet` type in my latest commit.

**Pull request merged**